File: 127_topological-sorting/topological-sorting.cpp
/*
@Copyright:LintCode
@Author:   heiyanbin
@Problem:  http://www.lintcode.com/problem/topological-sorting
@Language: C++
@Datetime: 15-01-17 03:33
*/

/**
 * Definition for Directed graph.
 * struct DirectedGraphNode {
 *     int label;
 *     vector<DirectedGraphNode *> neighbors;
 *     DirectedGraphNode(int x) : label(x) {};
 * };
 */
class Solution {
public:
    /**
     * @param graph: A list of Directed graph node
     * @return: Any topological order for the given graph.
     */
    vector<DirectedGraphNode*> topSort(vector<DirectedGraphNode*> graph) {
        for (auto node : graph)
            if (visited.find(node) == visited.end())
                dfs(node);
        return vector<DirectedGraphNode*>(res.rbegin(),res.rend());
    }
private:
    void dfs(DirectedGraphNode* start) {
        visited.insert(start);
        for (auto link : start -> neighbors)
            if (visited.find(link) == visited.end())
                dfs(link);
        res.push_back(start);
    }
    unordered_set<DirectedGraphNode*> visited;
    vector<DirectedGraphNode*> res;
};

File: 127_topological-sorting/topological-sorting.py
# coding:utf-8
'''
@Copyright:LintCode
@Author:   heiyanbin
@Problem:  http://www.lintcode.com/problem/topological-sorting
@Language: Python
@Datetime: 15-05-07 23:03
'''

# Definition for a Directed graph node
# class DirectedGraphNode:
#     def __init__(self, x):
#         self.label = x
#         self.neighbors = []

class Solution:
    """
    @param graph: A list of Directed graph node
    @return: A list of integer
    """
    def topSort(self, graph):
        # write your code here
        map = {}
        def mapping(node):
            if node == None: return
            map[node] = DirectedGraphNode(node.label)
            for link in node.neighbors:
                if link not in map:
                    mapping(link)
        for vertex in graph:
            if vertex not in map:
                mapping(vertex)
        
        for k, v in map.iteritems():
            for link in k.neighbors:
                map[link].neighbors.append(v)
        
        marked, topologyOrder = set([]), []
        def dfs(node):
            marked.add(node)
            for link in node.neighbors:
                if link not in marked:
                    dfs(link)
            topologyOrder.append(node)
            
        for vertex in map.itervalues():
            if vertex not in marked:
                dfs(vertex)
        
        return topologyOrder
                    
